### ISSUE TITLE
Header must provide a Content-Length property

### ISSUE THREAD

**Issue by jcagarcia** [Shopify/vscode-ruby-lsp#850]
### Operating System

OSX Sonoma 14.0

### Ruby version

3.2.2

### Project has a bundle

- [X] Has bundle

### Ruby version manager being used

rvm

### Description

After opening a Ruby project using VSCode, I'm getting the following error when trying to initialize the Ruby LSP extension

```
Ruby LSP> Skipping custom bundle setup since /Users/Juancarlos.Garcia/code/my-project/.ruby-lsp/Gemfile.lock already exists and is up to date
Ruby LSP> Running bundle install for the custom bundle. This may take a while...
The Gemfile's dependencies are satisfied
Fetching gem metadata from https://rubygems.org/.......
Resolving dependencies...
Bundler attempted to update ruby-lsp but its version stayed the same
Bundler attempted to update debug but its version stayed the same
Bundle updated!
Starting Ruby LSP...
[Error - 20:47:53] Client Ruby LSP: connection to server is erroring.
Header must provide a Content-Length property.
{"resolving dependencies...\ncontent-length":"1539"}
Shutting down server.
[Error - 20:47:53] Stopping server failed
Error: Client is not running and can't be stopped. It's current state is: starting
	at $d.shutdown (/Users/Juancarlos.Garcia/.vscode/extensions/shopify.ruby-lsp-0.4.10/out/extension.js:41:8468)
	at $d.stop (/Users/Juancarlos.Garcia/.vscode/extensions/shopify.ruby-lsp-0.4.10/out/extension.js:41:8047)
	at $d.stop (/Users/Juancarlos.Garcia/.vscode/extensions/shopify.ruby-lsp-0.4.10/out/extension.js:44:38188)
	at $d.handleConnectionError (/Users/Juancarlos.Garcia/.vscode/extensions/shopify.ruby-lsp-0.4.10/out/extension.js:43:63)
	at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
[Error - 20:47:53] Client Ruby LSP: connection to server is erroring.
Header must provide a Content-Length property.
{"{\"id\"":"0,\"result\":{\"capabilities\":{\"positionEncoding\":\"utf-16\",\"textDocumentSync\":{\"openClose\":true,\"change\":2},\"completionProvider\":{\"triggerCharacters\":[\"/\",\"A\",\"B\",\"C\",\"D\",\"E\",\"F\",\"G\",\"H\",\"I\",\"J\",\"K\",\"L\",\"M\",\"N\",\"O\",\"P\",\"Q\",\"R\",\"S\",\"T\",\"U\",\"V\",\"W\",\"X\",\"Y\",\"Z\"],\"completionItem\":{\"labelDetailsSupport\":true}},\"hoverProvider\":{},\"definitionProvider\":true,\"documentHighlightProvider\":true,\"documentSymbolProvider\":{\"symbolKind\":{\"value_set\":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]},\"hierarchicalDocumentSymbolSupport\":true},\"codeActionProvider\":{\"resolveProvider\":true},\"codeLensProvider\":{},\"documentLinkProvider\":{},\"documentFormattingProvider\":true,\"documentOnTypeFormattingProvider\":{\"firstTriggerCharacter\":\"{\",\"moreTriggerCharacter\":[\"\\n\",\"|\"]},\"foldingRangeProvider\":{\"lineFoldingOnly\":true},\"selectionRangeProvider\":true,\"semanticTokensProvider\":{\"documentSelector\":{\"scheme\":\"file\",\"language\":\"ruby\"},\"legend\":{\"tokenTypes\":[\"namespace\",\"type\",\"class\",\"enum\",\"interface\",\"struct\",\"typeParameter\",\"parameter\",\"variable\",\"property\",\"enumMember\",\"event\",\"function\",\"method\",\"macro\",\"keyword\",\"modifier\",\"comment\",\"string\",\"number\",\"regexp\",\"operator\",\"decorator\"],\"tokenModifiers\":[\"declaration\",\"definition\",\"readonly\",\"static\",\"deprecated\",\"abstract\",\"async\",\"modification\",\"documentation\",\"default_library\"]},\"range\":true,\"full\":{\"delta\":false}},\"inlayHintProvider\":{},\"diagnosticProvider\":{\"interFileDependencies\":false,\"workspaceDiagnostics\":false},\"workspaceSymbolProvider\":true}},\"jsonrpc\":\"2.0\"}Content-Length: 242"}
Shutting down server.
[Error - 20:47:53] Stopping server failed
Error: Client is not running and can't be stopped. It's current state is: starting
	at $d.shutdown (/Users/Juancarlos.Garcia/.vscode/extensions/shopify.ruby-lsp-0.4.10/out/extension.js:41:8468)
	at $d.stop (/Users/Juancarlos.Garcia/.vscode/extensions/shopify.ruby-lsp-0.4.10/out/extension.js:41:8047)
	at $d.stop (/Users/Juancarlos.Garcia/.vscode/extensions/shopify.ruby-lsp-0.4.10/out/extension.js:44:38188)
	at $d.handleConnectionError (/Users/Juancarlos.Garcia/.vscode/extensions/shopify.ruby-lsp-0.4.10/out/extension.js:43:63)
	at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```

**Comment by vinistock** [comment]
Thank you for the bug report!

It seems that some bundler output is getting mixed with the Ruby LSP's responses in STDOUT based on this line, which should only have included the content length.
```
{"resolving dependencies...\ncontent-length":"1539"}
```

Do you have any bundler configuration by any chance? I'm not sure how that bundler output ended up mixed in there - especially given that we [redirect all bundler output to STDERR on the server](https://github.com/Shopify/ruby-lsp/blob/3ba35a5b25b2ade35c446a184d7906b68d1247b4/lib/ruby_lsp/setup_bundler.rb#L201).

**Comment by vinistock** [comment]
Folks, we'll need more information to be able to understand what's going on. If someone who can reproduce the issue help debug it would help a lot.

The following script will run our custom bundle logic and capture anything that got printed to STDOUT. We need to understand what's being printed to STDOUT, because that's how the output would get mixed into responses that we send back to the editor.

```ruby
# Require the Ruby LSP's custom bundle logic
require "ruby_lsp/setup_bundler"

# Require StringIO
require "stringio"

# Save the current $stdout in a variable
current_stdout = $stdout

# Set $stdout to be a StringIO so that we can capture it
$stdout = StringIO.new

# Run the Ruby LSP's custom bundle logic
_bundle_gemfile, _bundle_path = RubyLsp::SetupBundler.new(Dir.pwd, **{}).setup!

# Get everything that was printed to STDOUT
content = $stdout.string

# Restore the original STDOUT
$stdout = current_stdout

# Show contents that were printed to STDOUT
puts "*" * 80
puts "STDOUT"
puts content
```

**Comment by jderose9** [comment]
@vinistock Happy to help but I don't understand what context this is supposed to run in? In the <project>/.ruby-lsp directory? Or replace something in the extension code?

**Comment by jderose9** [comment]
I don't get any STDOUT:

```
Ruby LSP> Skipping custom bundle setup since /Users/jason/code/my-project/.ruby-lsp/Gemfile.lock already exists and is up to date
Ruby LSP> Running bundle install for the custom bundle. This may take a while...
Ruby LSP> Command: (bundle check || bundle install) 1>&2
The Gemfile's dependencies are satisfied
********************************************************************************
STDOUT

```

**Comment by vinistock** [comment]
Interesting...still not sure how other output is getting mixed up. Do you have any bundler configurations?

@bricedurand are you also using RVM? Trying to understand if this problem is somehow related to RVM specifically.

**Comment by jderose9** [comment]
@vinistock I didn't dig into the extension code to see what it's doing, but reading just the console output, it seems like it gets past the bundle configuration. It says it's skipping the bundle setup as that has already been completed. So I'm not sure this is bundler output. It appears that the extension has both a client and server component, and the message is output when the extension is attempting to connect to the server.

**Comment by vinistock** [comment]
Based on the backtrace posted, it looks like bundler output is getting mixed into the responses we send back to the editor somehow.

```
{"resolving dependencies...\ncontent-length":"1539"}
```

That line should only contain the `content-length`, so that the server knows how many bytes to read from the STDIN pipe. I have no idea where the `resolving dependencies...` part is coming from or how that's getting mixed up. Isn't that from Bundler? Does RVM print things to stdout when you run [rvm-auto-ruby](https://github.com/Shopify/vscode-ruby-lsp/blob/f105bef0c5a8b74707c153576975b08da05bceb1/src/ruby.ts#L103)?

**Comment by marciotoshio** [comment]
Hi, I was facing the same issue and I can confirm that 0.4.14 is working. 👍🏽 
![Screenshot 2023-11-12 at 20 29 37](https://github.com/Shopify/vscode-ruby-lsp/assets/454861/9547f957-6000-4016-a012-d6eae07b5d2f)

**Comment by flooguu** [comment]
> I released https://github.com/Shopify/vscode-ruby-lsp/releases/tag/v0.4.14 as a prerelease with the fix. If someone could please install the prerelease version and let us know if the fix worked, I'd really appreciate it.

Works for me, thanks!

**Comment by vinistock** [comment]
Happy to hear that it worked. I'll close this for now, but please let us know if you find any other issues.

**Comment by bricedurand** [comment]
Hi there, unfortunately the pre release v0.4.14 still isn't working for me.

Here is the output I get 
```
Ruby LSP> Trying to activate Ruby environment with command: /bin/zsh -ic 'rvm-auto-ruby -rjson -e "STDERR.printf(%{RUBY_ENV_ACTIVATE%sRUBY_ENV_ACTIVATE}, JSON.dump(ENV.to_h))"' inside directory: /Users/brice/Documents/dev/collectif-objets
Ruby LSP> Skipping custom bundle setup since /Users/brice/Documents/dev/collectif-objets/.ruby-lsp/Gemfile.lock already exists and is up to date
Ruby LSP> Running bundle install for the custom bundle. This may take a while...
Ruby LSP> Command: (bundle check || bundle install) 1>&2
Found no changes, using resolution from the lockfile
The Gemfile's dependencies are satisfied
Starting Ruby LSP v0.12.3...
[Error - 09:31:31] Client Ruby LSP: connection to server is erroring.
Header must provide a Content-Length property.
{"found no changes, using resolution from the lockfile\ncontent-length":"1539"}
Shutting down server.
[Error - 09:31:31] Stopping server failed
Error: Client is not running and can't be stopped. It's current state is: starting
	at Hd.shutdown (/Users/brice/.vscode/extensions/shopify.ruby-lsp-0.4.14/out/extension.js:41:8468)
	at Hd.stop (/Users/brice/.vscode/extensions/shopify.ruby-lsp-0.4.14/out/extension.js:41:8047)
	at Hd.stop (/Users/brice/.vscode/extensions/shopify.ruby-lsp-0.4.14/out/extension.js:44:38188)
	at Hd.handleConnectionError (/Users/brice/.vscode/extensions/shopify.ruby-lsp-0.4.14/out/extension.js:43:63)
	at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```

**Comment by vinistock** [comment]
@bricedurand thank you for the backtrace, I think I figured out the problem now. You must have Bundler's log level configured to debug, because you're getting debug messages like [this one](https://github.com/rubygems/rubygems/blob/8a6b180d0ae5f14239506ac0469933ca062a8b47/bundler/lib/bundler/definition.rb#L292) printed to stdout (which messes up client/server communication).

I put up a PR on the server to override the log level to `silent`, which I believe will fix the issue. You can also verify if this is the case by changing your log level configuration.

**Comment by vinistock** [comment]
You have `found no changes, using resolution from the lockfile` in your backtrace.

There's only one place in the entire rubygems codebase where that string is printed, which is the point in the code that I linked to.

It uses `Bundler.ui.debug`, which [only prints](https://github.com/rubygems/rubygems/blob/8a6b180d0ae5f14239506ac0469933ca062a8b47/bundler/lib/bundler/ui/shell.rb#L50) if the [level](https://github.com/rubygems/rubygems/blob/8a6b180d0ae5f14239506ac0469933ca062a8b47/bundler/lib/bundler/ui/shell.rb#L71) is set to debug. The only way that string is getting printed is if the log level is set to debug, I don't see any alternative.

That said, there are multiple ways to configure it, one of them being the [verbose flag](https://github.com/rubygems/rubygems/blob/8a6b180d0ae5f14239506ac0469933ca062a8b47/bundler/lib/bundler/cli.rb#L79). The resolver also accepts some [environment variables](https://github.com/rubygems/rubygems/blob/8a6b180d0ae5f14239506ac0469933ca062a8b47/bundler/lib/bundler/resolver.rb#L187)

**Comment by eneagoe** [comment]
If it helps anyone, I had the same issue resolved by updating bundler to `2.4.22`

**Comment by e-smirnov735** [comment]
> I have the same issue after migrating from the Ruby and VSCode Ruby extensions.
> 
> I did remove the old configuration (starting with "ruby.") and followed the Troubleshooting section.



> I released https://github.com/Shopify/vscode-ruby-lsp/releases/tag/v0.4.14 as a prerelease with the fix. If someone could please install the prerelease version and let us know if the fix worked, I'd really appreciate it.

Works!
